### PR TITLE
minor update to the repo to ensure running in all machines

### PR DESCRIPTION
**Pull request by MaxHao56**
1. Graph fusion layer with attention mask = None
2. sweeper = basics
3. dataset bool

Optional Changes on the side. YAML file in the configs for giga pretrain needs to have match to your directory in you local or cluster machine

### CODE REVIEW

File: src/configs/dataset/giga_pretrain.yaml [hunk 1]
```diff
@@ -3,7 +3,9 @@ defaults:
 
 dataset:
   _target_: tasks.contrastive_learning.contrastive_dataset.ContrastivePreTrainingDataset
-  root: ${paths.root_dir}/tasks/contrastive_learning/giga_raw_files/
+  # root: ${paths.root_dir}/tasks/contrastive_learning/giga_raw_files/
+  root: /u201/mhao/MDT-2-1/src/tasks/contrastive_learning/giga_raw_files/
```
Comment (liamhebert) [inline src/configs/dataset/giga_pretrain.yaml]: Can you expand why this is need from the original version? `paths.root_dir` should resolve to be `/u201/mhao/MDT-2-1/src`

File: src/configs/dataset/hateful_discussions.yaml [hunk 2]
```diff
@@ -3,7 +3,8 @@ defaults:
 
 dataset:
   _target_: tasks.hateful_discussions.hateful_discussions.HatefulDiscussions
-  root: ${paths.root_dir}/tasks/hateful_discussions/raw_files/
+  # root: ${paths.root_dir}/tasks/hateful_discussions/raw_files/
+  root: /u201/mhao/MDT-2-1/src/tasks/hateful_discussions/raw_files
```
Comment (liamhebert) [inline src/configs/dataset/hateful_discussions.yaml]: Ditto here on why this is needed. It does make it less portable

**Review by liamhebert — CHANGES_REQUESTED** [review]
Few minor comments on `root_dir`, which when resolved are good to submit :)

File: src/configs/dataset/giga_pretrain.yaml (hunk 1 above)
Comment (liamhebert) [reply src/configs/dataset/giga_pretrain.yaml]: The change in root dir is also failing some tests: 
```
FAILED test_configs.py::test_train_config[graphormer_hatefuldiscussions2] - hydra.errors.InstantiationException: Error in call to target 'tasks.hateful_discussions.hateful_discussions.HatefulDiscussions':
FileNotFoundError(2, 'No such file or directory')
```

File: logs/train/runs/2025-10-24_04-31-01/.hydra/config.yaml
Comment (liamhebert) [inline logs/train/runs/2025-10-24_04-31-01/.hydra/config.yaml:1]: We probably don't want to commit these log files, right?

File: src/checkpt.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#  Allow OmegaConf globals (required for Lightning checkpoints)
+torch.serialization.add_safe_globals(
+    [
+        omegaconf.listconfig.ListConfig,
+        omegaconf.dictconfig.DictConfig,
+    ]
+)
+
+#  Force unsafe load explicitly (trusted file, your own)
+ckpt = torch.load(ckpt_path, map_location="cpu", weights_only=False)
```
Comment (liamhebert) [inline src/checkpt.py:41]: Not sure if this file is still needed since we swapped to using Hydra inits. If it is, can we replace this to instead load hydra configs and print them that way? Hydra can read in yamls and print the full structure like a json, which might achieve the same goal as this. (or, you know, we could look at the yaml directly :) ). 

If this isn't needed, suggest deleting this file.

File: src/debug_local.sh [hunk 4]
```diff
@@ -1,4 +1,9 @@
 #!/bin/sh
 export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
 
-srun python train.py experiment=giga_pretrain_roberta_anchor logger=wandb env=all hparams_search=pretrain hydra/launcher=basic hydra.sweeper.n_trials=1 trainer=gpu trainer.max_epochs=3 trainer.min_epochs=3 dataset.dataset.debug=40
+# srun python train.py experiment=giga_pretrain_roberta_anchor logger=wandb env=all hparams_search=pretrain hydra/launcher=basic hydra.sweeper=basic trainer=gpu trainer.max_epochs=3 trainer.min_epochs=3 dataset.dataset.debug=40
+srun python train.py experiment=giga_pretrain_roberta_anchor logger=wandb env=all env=all hydra/launcher=submitit_slurm trainer=gpu trainer.max_epochs=3 trainer.min_epochs=3 dataset.dataset.debug=5
```
Comment (liamhebert) [inline src/debug_local.sh]: can we revert these edits?

File: src/run_graphormer_hateful_discussions.sh [hunk 5]
```diff
@@ -12,5 +12,7 @@
 
 export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
 export TOKENIZERS_PARALLELISM=false
+export HYDRA_FULL_ERROR=1
 
-python train.py experiment=graphormer_hatefuldiscussions2 trainer=gpu logger=wandb logger.wandb.project=graphormer_hd2 env=all
+# python train.py experiment=graphormer_hatefuldiscussions2 trainer=gpu logger=wandb logger.wandb.project=graphormer_hd2 env=all
```
Comment (liamhebert) [inline src/run_graphormer_hateful_discussions.sh]: likewise, can these edits be reverted?

File: src/test_data.embeddings.npy
Comment (liamhebert) [inline src/test_data.embeddings.npy:1]: revert?

File: src/test_data.json
Comment (liamhebert) [inline src/test_data.json:1]: likewise, revert?

File: src/test_embedding.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+from hydra.utils import instantiate
+from transformers import AutoTokenizer, AutoImageProcessor
+from PIL import Image
+from data.types import GraphFeatures, TextFeatures, ImageFeatures
+from data import collator_utils
+from tasks import dataset_utils as dut
+
+# ---- Pathing ----
+ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+sys.path.extend([ROOT, os.path.join(ROOT, "src")])
```
Comment (liamhebert) [inline src/test_embedding.py]: Suggest using https://github.com/liamhebert/MDT-2/blob/5f50b4abbf32d626f9695617751a38b778e3f224/src/__init__.py#L5-L7 
which would achieve the same goal and might be a bit more clean

File: src/test_embedding.py [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+------
+python src/test_embedding.py \
+  --json src/test_data.json \
+  --ckpt src/logs/pretrain/runs/2025-10-23_22-41-28/checkpoints/last.ckpt \
+  --device cuda \
+
+"""
+
+# the batch size command is removed as it actually valids the parameters,
+# need to see why
```
Comment (liamhebert) [inline src/test_embedding.py]: Comment still relevant?

File: src/test_embedding.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+from tasks import dataset_utils as dut
+
+# ---- Pathing ----
+ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+sys.path.extend([ROOT, os.path.join(ROOT, "src")])
+
+
+#  Convert a tree jsn to feature dicts, the graph
+#  watch: Similar to data.dataset_utils.tree_to_graph_features but simplified
+def tree_to_graph_features(tree, text_tokenizer, image_processor, root_dir="."):
```
Comment (liamhebert) [inline src/test_embedding.py:39]: Please add type annotations here. 
(ie tree is a dict[str, Any], text_tokenizer is a transformers.PreTrainedTokenizer (https://huggingface.co/docs/transformers/v4.57.1/en/main_classes/tokenizer#transformers.PreTrainedTokenizer) and image_processor is something...

File: src/test_embedding.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    }
+
+    def traverse(node, parent_id=None):
+        if parent_id is None:
+            parent_id = node["id"]
+        flat["id"].append(node["id"])
+        flat["parent_id"].append(parent_id)
+        flat["text"].append(dut.clean_text(node.get("body", "")))
+        flat["images"].append(
+            node.get("images", [None])[0] if node.get("images") else None
```
Comment (liamhebert) [inline src/test_embedding.py:58]: this can be simplied to just `node.get("images")` (https://docs.python.org/3.8/library/stdtypes.html#dict.get) 
Calling `node.get("images")` will return `None` if the key is not present.

File: src/test_embedding.py [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        )
+        flat["rotary_position"].append(node.get("rotary_position", [0, 0]))
```
Comment (liamhebert) [inline src/test_embedding.py]: hm, I'm guessing the else condition `[0, 0]` should never get hit right? If we did, something would be pretty wrong. Can we replace this to `assert "rotary_position" in node` before getting the value?

File: src/test_embedding.py [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+        for c in node.get("tree", []):
+            traverse(c, node["id"])
+
+    traverse(tree)
+
+    # --- Build graph tensors ---
+    n = len(flat["id"])
+    id_map = {nid: i for i, nid in enumerate(flat["id"])}
```
Comment (liamhebert) [inline src/test_embedding.py]: nit: replace `nid` with `node_id`

File: src/test_embedding.py (hunk 9 above)
Comment (liamhebert) [reply src/test_embedding.py:58]: On second thought, it seems like the fact that `images` being a list complicates things, ugh. either way,  `node.get("images", [None])[0]` will work too.

File: src/test_embedding.py [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+sys.path.extend([ROOT, os.path.join(ROOT, "src")])
+
+
+#  Convert a tree jsn to feature dicts, the graph
+#  watch: Similar to data.dataset_utils.tree_to_graph_features but simplified
+def tree_to_graph_features(tree, text_tokenizer, image_processor, root_dir="."):
+    """Flatten a discussion tree into tensors for model input."""
+    dut.compute_relative_distance(tree)
+
+    flat = {
```
Comment (liamhebert) [inline src/test_embedding.py:43]: We are missing a few parameters here. Namely, `is_root` and `distances`. These should be already attributes of node (I think). 

https://github.com/liamhebert/MDT-2/blob/5f50b4abbf32d626f9695617751a38b778e3f224/src/tasks/dataset.py#L860-L867

File: src/test_embedding.py [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+        [img is not None for img in flat["images"]], dtype=torch.bool
+    )
+
+    graph_features = {
+        "edge_index": edges,
+        "in_degree": in_degree,
+        "out_degree": out_degree,
+        "attn_bias": torch.zeros((n, n)),
+        "distance": torch.zeros((n, n, 2)),
+        "distance_index": torch.zeros((n, n), dtype=torch.int16),
```
Comment (liamhebert) [inline src/test_embedding.py:94]: This doesn't feel right... can you double check whether there should be a value here?

File: src/test_embedding.py [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+        padding="max_length",
+        truncation=True,
+        return_tensors="pt",
+        max_length=256,
+    )
+    text_features = {
+        "input_ids": tokenized_text["input_ids"],
+        "attention_mask": tokenized_text["attention_mask"],
+        "token_type_ids": tokenized_text.get(
+            "token_type_ids", torch.zeros_like(tokenized_text["input_ids"])
```
Comment (liamhebert) [inline src/test_embedding.py:113]: likewise, we should be careful here

**Review by liamhebert — CHANGES_REQUESTED** [review]
Still need to review EmbeddingGenerator, more soon.

File: logs/train/runs/2025-10-24_04-31-01/.hydra/config.yaml
Comment (MaxHao56) [reply logs/train/runs/2025-10-24_04-31-01/.hydra/config.yaml:1]: I have deleted it them.

File: src/checkpt.py (hunk 3 above)
Comment (MaxHao56) [reply src/checkpt.py:41]: Sounds Good.

File: src/debug_local.sh (hunk 4 above)
Comment (MaxHao56) [reply src/debug_local.sh]: completed

File: src/run_graphormer_hateful_discussions.sh (hunk 5 above)
Comment (MaxHao56) [reply src/run_graphormer_hateful_discussions.sh]: completed

File: src/test_data.embeddings.npy
Comment (MaxHao56) [reply src/test_data.embeddings.npy:1]: completed

File: src/test_data.json
Comment (MaxHao56) [reply src/test_data.json:1]: completed

File: src/test_embedding.py (hunk 6 above)
Comment (MaxHao56) [reply src/test_embedding.py]: completed

File: src/test_embedding.py (hunk 7 above)
Comment (MaxHao56) [reply src/test_embedding.py]: completed
